### PR TITLE
Add LessOrEqual and GreaterOrEqual Ops.

### PR DESCRIPTION
**Pull request by zrphercule**
Right now, we have:

tf.math.less_equal() tf.math.greater_equal()
c2.LE() c2.GE()
mxnet::symbol.__le__() mxnet::symbol.__ge__()
np.less_equal() np.greater_equal()

Therefore it is neccesary to add two ops in ONNX as well.

### CODE REVIEW

**Comment by zrphercule** [comment]
I will generate and upload docs once it is approved.

**Comment by houseroad** [comment]
@linkerzhang @ebarsoum @spandantiwari @pk-g do you mind to take a look?

File: onnx/defs/logical/defs.cc [hunk 1]
```diff
@@ -121,6 +121,34 @@ ONNX_OPERATOR_SET_SCHEMA(
...
+	OpSchema()
+		.FillUsing(BinaryLogicDocGenerator("greater or equal"))
+		.TypeConstraint(
+			"T",
+			OpSchema::all_numeric_types(),
+			"Constarins input to float tensors.")
+		.TypeConstraint(
+			"T1",
+			{"tensor(bool)"},
+			"Constrains output to boolean tensor."));
```
Comment (spandantiwari) [inline onnx/defs/logical/defs.cc]: Maybe consider adding type and shape inference function as they are increasingly being used by backends for model validation.

File: onnx/defs/logical/defs.cc [hunk 2]
```diff
@@ -138,4 +166,6 @@ ONNX_OPERATOR_SET_SCHEMA(
             "Constrains input/output to boolean tensors.")
         .TypeAndShapeInferenceFunction(unaryLogicalOpInference));
 
+
```
Comment (spandantiwari) [inline onnx/defs/logical/defs.cc]: line break - intentional?

File: onnx/defs/logical/defs.cc [hunk 3]
```diff
@@ -121,6 +121,34 @@ ONNX_OPERATOR_SET_SCHEMA(
...
 
+ONNX_OPERATOR_SET_SCHEMA(
+	LessOrEqual,
+	9,
+	OpSchema()
+		.FillUsing(BinaryLogicDocGenerator("less or equal"))
+		.TypeConstraint(
+			"T",
+			OpSchema::all_numeric_types(),
+			"Constarins input to float tensors.")
```
Comment (spandantiwari) [inline onnx/defs/logical/defs.cc]: The comment does not match the types allowed. Also nit: spelling for "constrains" .

File: onnx/defs/logical/defs.cc [hunk 4]
```diff
@@ -121,6 +121,34 @@ ONNX_OPERATOR_SET_SCHEMA(
...
+
+ONNX_OPERATOR_SET_SCHEMA(
+	GreaterOrEqual,
+	9,
+	OpSchema()
+		.FillUsing(BinaryLogicDocGenerator("greater or equal"))
+		.TypeConstraint(
+			"T",
+			OpSchema::all_numeric_types(),
+			"Constarins input to float tensors.")
```
Comment (spandantiwari) [inline onnx/defs/logical/defs.cc]: Same comment as above.

File: onnx/defs/logical/defs.cc (hunk 1 above)
Comment (spandantiwari) [inline onnx/defs/logical/defs.cc]: nit: spelling

File: onnx/backend/test/case/model/greater_or_equal.py [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+                                                       onnx.TensorProto.FLOAT,
+                                                       x.shape),
+                    onnx.helper.make_tensor_value_info('y',
+                                                       onnx.TensorProto.FLOAT,
+                                                       y.shape)],
+            outputs=[onnx.helper.make_tensor_value_info('z',
+                                                        onnx.TensorProto.FLOAT,
+                                                        z.shape)])
+        model = onnx.helper.make_model(graph, producer_name='backend-test')
+        expect(model, inputs=[x], outputs=[y],
```
Comment (spandantiwari) [inline onnx/backend/test/case/model/greater_or_equal.py]: Great to see both node and model tests for these ops. :-)

File: docs/Changelog.md [hunk 6]
```diff
@@ -8855,6 +8855,42 @@ This version of the operator has been available since version 9 of the default O
 <dd>Constrains output to boolean tensor.</dd>
 </dl>
 
+### <a name="GreaterOrEqual-9"></a>**GreaterOrEqual-9**</a>
+
+  Returns the tensor resulted from performing the `greater or equal` logical operation
```
Comment (spandantiwari) [inline docs/Changelog.md]: I suggest slight rephrasing: "Returns tensor resulting from..."

**Comment by spandantiwari** [comment]
@zrphercule @houseroad - The spec looks good to me. 

I do have a high-level question. Do we want to introduce these as separate ops given that we can obtain the effect by simple negation of existing ops. E.g. `LessOrEqual` can be implemented by `Greater` followed by `Not`. I feel that it is OK to include them in the spec because they are common logical operators and have clear design. At the same time, such additions add to the number of supported ops. 

For e.g., we have an `Equal` op and a `Not` op. Does the spec have place for a `NotEqual` op?

**Comment by shinh** [comment]
`Greater` and `Not` is wrong for NaNs?

```
>>> nan = np.array(float('nan'))
>>> nan <= nan
False
>>> not (nan > nan)
True
```

File: onnx/defs/logical/defs.cc (hunk 3 above)
Comment (zrphercule) [reply onnx/defs/logical/defs.cc]: Yeah, and I just noticed the schema of Greater and Less were also wrong here.

**Comment by zrphercule** [comment]
@spandantiwari Thanks for your review! I have two things would like to clear:

1. For these two ops, we have a need to add them because many backend frameworks have them in their op list, and may have some special optimize for them. Also, as @shinh mentioned, sometimes "not greater" is not fully equal to "less or equal" , e.g. the two numbers are not comparable. As for NotEqual, if it meet the these two conditions, I guess it is also okay to have it in ONNX?

2. I think you mentioned a very good point: when should we add one op as a "function", and when should we add it as a normal operator. @houseroad  @bddppq  and I think it might be better to give both options to frameworks and let themselves decide. 
For example, in <= case, we create a standard op LessOrEqual, meanwhile we add an optional attribute "isFunc" in it.
If isFunc is given, then we will treat LessOrEqual as a function instead of a standard operator, and dismantle it to be "Not Greater" in our exported model. This solution is more flexible, we can have a standard op while have one or even more pre-setting ways of translating it to a function. We can also let backend decide how to dismantle an op on their own by providing a subgraph as optional Attribute. Right now, our function design uses a totally different workflow, and one op can only be a function OR a normal op.

What is your oppinion? Thanks!

File: docs/Changelog.md (hunk 6 above)
Comment (zrphercule) [reply docs/Changelog.md]: ok sure, this will update the docs of all logical ops, and I think it is fine =w=

**Comment by spandantiwari** [comment]
@zrphercule - I understand and agree with your points about being able to optimize better when `LessOrEqual` is an op in itself and also about it producing different values for edge cases with `NaNs`. Regarding the second point - while I agree that we should let the frameworks, my point was about the number of new ops that this introduces (that's why I included the example for `NotEqual`) in the ONNX spec. Let's say that we let the framework decide which way to export, the ONNX spec still needs to have the `LessOrEqual` op in case the framework wants to export it. And this increases the number of ops. 

In this particular case I am fine with including these ops.But I have heard concerns from others about increasing number ops and that's why I would love to hear their thoughts.

**Comment by prasanthpul** [comment]
One benefit of adding as a function is that runtimes can run models that use it right away, even before they have implemented an optimized implementation of that function

**Pull request closed without merging**